source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -262,35 +289,33 @@ bool Filter::handleResponseBodyResponse(const BodyResponse& response) {
       MutationUtils::applyCommonBodyResponse(response, data);
     });
     response_state_ = FilterState::Idle;
+    response_message_timer_->disableTimer();
     encoder_callbacks_->continueEncoding();
     return true;
   }
   return false;
 }
 
-void Filter::handleImmediateResponse(const ImmediateResponse& response) {
-  // We don't want to process any more stream messages after this.
-  // Close the stream before sending because "sendLocalResponse" may trigger
-  // additional calls to this filter.
-  processing_complete_ = true;
-  sendImmediateResponse(response);
-}
-
 void Filter::onGrpcError(Grpc::Status::GrpcStatus status) {
   ENVOY_LOG(debug, "Received gRPC error on stream: {}", status);
   stats_.streams_failed_.inc();
 
+  if (processing_complete_) {
+    return;
+  }
+
   if (config_->failureModeAllow()) {
     // Ignore this and treat as a successful close
     onGrpcClose();
     stats_.failure_mode_allowed_.inc();
 
   } else {
     processing_complete_ = true;
+    cancelTimers();
```

Comment:
Maybe a comment saying this gets canceled as part of `onGrpcClose` above: at first I was confused why we weren't canceling the timers in all cases

Reply:
I added a comment and also renamed the method "cleanUpTimers" to make clear that it's there to ensure that they don't fire any more when they'll no longer be wanted.